### PR TITLE
Fix printfn "%A" output for TypedArrrays.

### PR DESCRIPTION
**Pull request by fdcastel**
My suggestion to avoid another fright like #87 ;'-)

Uses [ArrayBuffer.isView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer/isView) to check for TypedArrays and the [second parameter of JSON.stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#The_replacer_parameter) to convert it to Array.

### CODE REVIEW

File: src/fable-js/fable-core.js [hunk 1]
```diff
@@ -315,7 +315,9 @@
           case "f": case "F": rep = rep.toFixed(precision || 6); break;
           case "g": case "G": rep = rep.toPrecision(precision); break;
           case "e": case "E": rep = rep.toExponential(precision); break;
-          case "A": rep = JSON.stringify(rep); break;
+          case "A": rep = JSON.stringify(rep, function(k, v) {
+                        return ArrayBuffer.isView(v) ? Array.from(v) : v;
```
Comment (alfonsogarciacaro) [inline src/fable-js/fable-core.js]: It's difficult to make `%A` work perfectly, but this is a very nice addition :) May I suggest to include other iterables too (lists, sets, maps)? Something like
```
return v && v[Symbol.iterator] && !Array.isArray(v) ? Array.from(v) : v;
```

Comment (fdcastel) [inline src/fable-js/fable-core.js]: Great! Better yet!

Comment (fdcastel) [inline src/fable-js/fable-core.js]: Let me do some tests here and I'll update the PR.

**Comment by fdcastel** [comment]
After giving some thoughts to this, I'm pushing what I think is a good compromise solution.

F# code used to test:

```
    let lst8  = [ 0uy; 1uy; 2uy; 3uy ]
    let lst16 = [ 0us; 1us; 2us; 3us ]
    let lst32 = [ 0; 1; 2; 3 ]
    let lstS =  [ "0"; "1"; "2"; "3" ]
        
    lst8  |> printfn "%A"
    lst16 |> printfn "%A"
    lst32 |> printfn "%A"
    lstS  |> printfn "%A"


    let arr8  = Array.ofList lst8
    let arr16 = Array.ofList lst16
    let arr32 = Array.ofList lst32
    let arrS =  Array.ofList lstS

    arr8  |> printfn "%A"
    arr16 |> printfn "%A"
    arr32 |> printfn "%A"
    arrS  |> printfn "%A"


    let seq8  = lst8  |> Seq.ofList
    let seq16 = lst16 |> Seq.ofList
    let seq32 = lst32 |> Seq.ofList
    let seqS  = lstS  |> Seq.ofList
        
    seq8  |> printfn "%A"
    seq16 |> printfn "%A"
    seq32 |> printfn "%A"
    seqS  |> printfn "%A"


    let set8  = lst8  |> Set.ofList
    let set16 = lst16 |> Set.ofList
    let set32 = lst32 |> Set.ofList
    let setS  = lstS  |> Set.ofList
        
    set8  |> printfn "%A"
    set16 |> printfn "%A"
    set32 |> printfn "%A"
    setS  |> printfn "%A"


    let fn = fun a -> (a, a.ToString())
    let map8  = Map.ofSeq (Seq.map fn arr8)
    let map16 = Map.ofSeq (Seq.map fn arr16)
    let map32 = Map.ofSeq (Seq.map fn arr32)
    let mapS  = Map.ofSeq (Seq.map fn arrS)

    map8  |> printfn "%A"
    map16 |> printfn "%A"
    map32 |> printfn "%A"
    mapS  |> printfn "%A"


    let fn = fun a -> (a, set8)
    let smap8  = Map.ofSeq (Seq.map fn arr8)
    let smap16 = Map.ofSeq (Seq.map fn arr16)
    let smap32 = Map.ofSeq (Seq.map fn arr32)
    let smapS  = Map.ofSeq (Seq.map fn arrS)

    smap8  |> printfn "%A"
    smap16 |> printfn "%A"
    smap32 |> printfn "%A"
    smapS  |> printfn "%A"
```

_(BTW: I didn't make a unit test for this PR since the results are "merely aesthetic". And are different of F# output, anyway)._

F# output:

```
[0uy; 1uy; 2uy; 3uy]
[0us; 1us; 2us; 3us]
[0; 1; 2; 3]
["0"; "1"; "2"; "3"]
[|0uy; 1uy; 2uy; 3uy|]
[|0us; 1us; 2us; 3us|]
[|0; 1; 2; 3|]
[|"0"; "1"; "2"; "3"|]
[0uy; 1uy; 2uy; 3uy]
[0us; 1us; 2us; 3us]
[0; 1; 2; 3]
["0"; "1"; "2"; "3"]
set [0uy; 1uy; 2uy; 3uy]
set [0us; 1us; 2us; 3us]
set [0; 1; 2; 3]
set ["0"; "1"; "2"; "3"]
map [(0uy, "0"); (1uy, "1"); (2uy, "2"); (3uy, "3")]
map [(0us, "0"); (1us, "1"); (2us, "2"); (3us, "3")]
map [(0, "0"); (1, "1"); (2, "2"); (3, "3")]
map [("0", "0"); ("1", "1"); ("2", "2"); ("3", "3")]
map
  [(0uy, set [0uy; 1uy; 2uy; 3uy]); (1uy, set [0uy; 1uy; 2uy; 3uy]);
   (2uy, set [0uy; 1uy; 2uy; 3uy]); (3uy, set [0uy; 1uy; 2uy; 3uy])]
map
  [(0us, set [0uy; 1uy; 2uy; 3uy]); (1us, set [0uy; 1uy; 2uy; 3uy]);
   (2us, set [0uy; 1uy; 2uy; 3uy]); (3us, set [0uy; 1uy; 2uy; 3uy])]
map
  [(0, set [0uy; 1uy; 2uy; 3uy]); (1, set [0uy; 1uy; 2uy; 3uy]);
   (2, set [0uy; 1uy; 2uy; 3uy]); (3, set [0uy; 1uy; 2uy; 3uy])]
map
  [("0", set [0uy; 1uy; 2uy; 3uy]); ("1", set [0uy; 1uy; 2uy; 3uy]);
   ("2", set [0uy; 1uy; 2uy; 3uy]); ("3", set [0uy; 1uy; 2uy; 3uy])]
```

Fable output (after e03699038d9f7837ca49b38ac74faa6a5e35a70c):

```
[0,1,2,3]
[0,1,2,3]
[0,1,2,3]
["0","1","2","3"]
[0,1,2,3]
[0,1,2,3]
[0,1,2,3]
["0","1","2","3"]
[0,1,2,3]
[0,1,2,3]
[0,1,2,3]
["0","1","2","3"]
set [0,1,2,3]
set [0,1,2,3]
set [0,1,2,3]
set ["0","1","2","3"]
map [[0,"0"],[1,"1"],[2,"2"],[3,"3"]]
map [[0,"0"],[1,"1"],[2,"2"],[3,"3"]]
map [[0,"0"],[1,"1"],[2,"2"],[3,"3"]]
map [["0","0"],["1","1"],["2","2"],["3","3"]]
map [[0,[0,1,2,3]],[1,[0,1,2,3]],[2,[0,1,2,3]],[3,[0,1,2,3]]]
map [[0,[0,1,2,3]],[1,[0,1,2,3]],[2,[0,1,2,3]],[3,[0,1,2,3]]]
map [[0,[0,1,2,3]],[1,[0,1,2,3]],[2,[0,1,2,3]],[3,[0,1,2,3]]]
map [["0",[0,1,2,3]],["1",[0,1,2,3]],["2",[0,1,2,3]],["3",[0,1,2,3]]]
```

_(blank lines removed)_

Note that in the last case the inner sets of map values are not displayed faithfully. It should be

`map [[0,set [0,1,2,3]],[1,set [0,1,2,3]],[2,set [0,1,2,3]],[3,set [0,1,2,3]]]`

however, I fear this can't be done with `JSON.stringify`. The only way I see to fix this would require a kind of "transclusion" which currently doesn't exists in `stringify` specs.

_(Of course I'll be glad if anyone could prove me wrong here ;) )_

An alternative implementation to make this could be

```
          case "A": 
            var f = function(k, v) {
                      return v && v[Symbol.iterator] 
                               && !Array.isArray(v) 
                               && !(typeof v === 'string' || v instanceof String)
                               ? ( (v instanceof Map) ? "map " :
                                   (v instanceof Set) ? "set " : "") + JSON.stringify(Array.from(v), f)
                               : v;
                    };
            rep = JSON.stringify(rep, f);
            break;
```

which outputs:

```
"[0,1,2,3]"
"[0,1,2,3]"
"[0,1,2,3]"
"[\"0\",\"1\",\"2\",\"3\"]"
"[0,1,2,3]"
"[0,1,2,3]"
"[0,1,2,3]"
["0","1","2","3"]
"[0,1,2,3]"
"[0,1,2,3]"
"[0,1,2,3]"
"[\"0\",\"1\",\"2\",\"3\"]"
"set [0,1,2,3]"
"set [0,1,2,3]"
"set [0,1,2,3]"
"set [\"0\",\"1\",\"2\",\"3\"]"
"map [[0,\"0\"],[1,\"1\"],[2,\"2\"],[3,\"3\"]]"
"map [[0,\"0\"],[1,\"1\"],[2,\"2\"],[3,\"3\"]]"
"map [[0,\"0\"],[1,\"1\"],[2,\"2\"],[3,\"3\"]]"
"map [[\"0\",\"0\"],[\"1\",\"1\"],[\"2\",\"2\"],[\"3\",\"3\"]]"
"map [[0,\"set [0,1,2,3]\"],[1,\"set [0,1,2,3]\"],[2,\"set [0,1,2,3]\"],[3,\"set [0,1,2,3]\"]]"
"map [[0,\"set [0,1,2,3]\"],[1,\"set [0,1,2,3]\"],[2,\"set [0,1,2,3]\"],[3,\"set [0,1,2,3]\"]]"
"map [[0,\"set [0,1,2,3]\"],[1,\"set [0,1,2,3]\"],[2,\"set [0,1,2,3]\"],[3,\"set [0,1,2,3]\"]]"
"map [[\"0\",\"set [0,1,2,3]\"],[\"1\",\"set [0,1,2,3]\"],[\"2\",\"set [0,1,2,3]\"],[\"3\",\"set [0,1,2,3]\"]]"
```

Yeah. Ugly as hell. :) The output of printfn "%A" should be somewhat "easy" on the eyes. Definitely not this.

**Comment by alfonsogarciacaro** [comment]
We're not forced to use `JSON.stringify`, I just found it was the easiest way to get some formatting for JS objects. Anyways, for simplicity I would merge the current version (unless you need something more specific). It shouldn't be very often that users put sets inside maps :)

**Pull request merged**